main.c

```diff
@@ -0,0 +1,200 @@
+#include <stdio.h>
+#include <stdlib.h>
+
+#include "canlib/can.h"
+#include "canlib/can_common.h"
+#include "canlib/pic18f26k83/pic18f26k83_can.h"
+#include "canlib/message_types.h"
+#include "canlib/util/timing_util.h"
+#include "canlib/util/can_tx_buffer.h"
+#include "canlib/pic18f26k83/pic18f26k83_timer.h"
+
+#include "arming.h"
+#include "altitude_parsing.h"
+#include "mcc_generated_files/mcc.h"
+
+#define _XTAL_FREQ 12000000 //12MHz
+
+static void can_msg_handler(const can_msg_t *msg);
+
+static enum ARM_STATE alt_1_arm_state = DISARMED;  //this should be ARMED for flight code
+static enum ARM_STATE alt_2_arm_state = DISARMED;
+
+// Memory pool for CAN transmit buffer
+uint8_t tx_pool[100];
+
+int main(int argc, char** argv) {
+    //init functions
+    osc_init();
+    timer0_init();    
+    output_init(); // initialize our outputs
+    uart1_rx_init(9600, _XTAL_FREQ); //initialize the UART module
+    // MCC generated functions
+    ADCC_Initialize(); 
+    FVR_Initialize();
+    
+    // enable global interupts
+    INTCON0bits.GIE = 1;
+    
+    /***********set up CAN**********/
+    
+    // Set up CAN TX
+    TRISC0 = 0;
+    RC0PPS = 0x33;
+
+    // Set up CAN RX
+    TRISC1 = 1;
+    ANSELC1 = 0;
+    CANRXPPS = 0x11;
+    
+    // set up CAN module
+    can_timing_t can_setup;
+    can_generate_timing_params(_XTAL_FREQ, &can_setup);
+
+    can_init(&can_setup, can_msg_handler);
+    // set up CAN tx buffer
+    txb_init(tx_pool, sizeof(tx_pool), can_send, can_send_rdy);
+    
+    
+    uint32_t last_millis = millis();
+    uint8_t on = 0;
+    
+    
+    /***************Main Loop***************/
+    while(1){
+        
+        if(millis() > last_millis + MAX_LOOP_TIME_DIFF_ms){
+            last_millis = millis();
+            
+            if(on){
+                WHITE_LED_OFF();
+                on = 0;
+            }
+            else{
+                WHITE_LED_ON();
+                on = 1;
+            }
+            
+            can_msg_t alt_1_arm_stat_msg;
+            build_arm_stat_msg(millis(), 
+                               1, 
+                               alt_1_arm_state,
+                               (uint16_t)ADCC_GetSingleConversion(A1_DROGUE_PIN)*3.72,
+                               (uint16_t)ADCC_GetSingleConversion(A1_MAIN_PIN)*3.72,
+                               &alt_1_arm_stat_msg);
+            txb_enqueue(&alt_1_arm_stat_msg);
+            
+            can_msg_t alt_2_arm_stat_msg;
+            build_arm_stat_msg(millis(), 
+                               2, 
+                               alt_2_arm_state,
+                               (uint16_t)(ADCC_GetSingleConversion(A2_DROGUE_PIN))*3.72,
+                               (uint16_t)(ADCC_GetSingleConversion(A2_MAIN_PIN))*3.72,
+                               &alt_2_arm_stat_msg);
```

Comment:
The magic number `3.72` appears in a lot of places with no explanation. It would be happier as a `#define` with a meaningful name.

Reply:
Also, are you expecting true analog voltages on the main/drogue pins, or just HIGH/LOW? If it's the latter, you can save yourself the ADC conversion and just read them as digital values (by reading from the relevant PORT). It won't make big difference since this board isn't doing a whole lot, but in principle you want to avoid unnecessary ADC conversions since they are quite slow.

Reply:
Some of these could be considered digital, but we can have the power to use the ADC, and the more exact data is useful.